File: training-data/r/factory_pattern.R
setClass("Circle", representation(radius = "numeric"))
setClass("Square", representation(side = "numeric"))

setGeneric("shape_area", function(shape) standardGeneric("shape_area"))
setMethod("shape_area", "Circle", function(shape) pi * shape@radius^2)
setMethod("shape_area", "Square", function(shape) shape@side^2)

shape_factory <- function(kind, param) {
  switch(kind,
    circle = new("Circle", radius = param),
    square = new("Square", side = param),
    stop("unknown shape: ", kind)
  )
}

circle <- shape_factory("circle", 2)
square <- shape_factory("square", 3)

print(shape_area(circle))
print(shape_area(square))

File: training-data/perl/factory_pattern.pl
use strict;
use warnings;

package Circle;

sub new {
    my ($class, $radius) = @_;
    return bless { radius => $radius }, $class;
}

sub area {
    my ($self) = @_;
    return 3.14159 * $self->{radius} ** 2;
}

package Square;

sub new {
    my ($class, $side) = @_;
    return bless { side => $side }, $class;
}

sub area {
    my ($self) = @_;
    return $self->{side} ** 2;
}

package ShapeFactory;

sub create {
    my ($class, $kind, @args) = @_;
    return Circle->new(@args) if $kind eq 'circle';
    return Square->new(@args) if $kind eq 'square';
    die "unknown shape: $kind\n";
}

package main;

for my $spec (['circle', 3], ['square', 4]) {
    my ($kind, @args) = @$spec;
    my $shape = ShapeFactory->create($kind, @args);
    printf "%s area: %.2f\n", $kind, $shape->area;
}
